### PR TITLE
[QoI] Improve diagnostics related to member calls

### PR DESCRIPTION
**Pull request by xedin**
When trying to diagnose problems related to calls where
function is represented by a member of nominal type let's
attempt to be more cautious while type-checking function
expression without it's arguments, because it could produce
unrelated diagnostics.

Resolves: rdar://problem/32551313, rdar://problem/28456467, rdar://problem/31671195

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
+  auto shouldTypeCheckFunction = [&](Expr *fnExpr) -> bool {
+    if (!isa<UnresolvedDotExpr>(fnExpr))
+      return true;
+
+    SmallVector<Type, 4> fnTypes;
+    CS->TC.getPossibleTypesOfExpressionWithoutApplying(
+        fnExpr, CS->DC, fnTypes, FreeTypeVariableBinding::UnresolvedType);
+
+    // Might be a structural problem related to the member itself.
+    if (fnTypes.empty())
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: Do you really want to check for `empty` explicitly? Not doing so results in returning true anyway.

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
     return true;
 
-  // Type check the function subexpression to resolve a type for it if
-  // possible.
-  auto fnExpr = typeCheckChildIndependently(callExpr->getFn());
-  if (!fnExpr)
+  auto originalFnType = CS->getType(callExpr->getFn());
+
+  // Check if there is a structural problem in the function expression
+  // itself, by type-checking it allowing unresolved variables. If
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: I had a hard time parsing this. Perhaps "Check if there is a structural problem in the function expression by performing type checking with the option to allow unresolved type variables."

File: lib/Sema/CSDiag.cpp [hunk 3]
```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
+  // that is going to produce a functional type with unresolved result
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: typo: functional type => function type

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
+  // let's not re-typecheck the function expression, because it might
+  // produce unrelated diagnostics due to lack of contextual information.
+  auto shouldTypeCheckFunction = [&](Expr *fnExpr) -> bool {
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: meta comment: `visitApplyExpr` is quite large. Rather than inlining lambdas like this, it probably makes sense to break them out into separate functions. It's already really hard to follow a lot of code due to the interspersing of specific decisions about how to diagnose problems and the support code for those decisions.

File: lib/Sema/CSDiag.cpp [hunk 5]
```diff
@@ -6246,6 +6282,35 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
 
   auto fnType = getFuncType(CS->getType(fnExpr));
 
+  // Let's see if this has to do with member vs. property error
+  // because sometimes when there is a member and a property declared
+  // on the nominal type with the same name type-checking function
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: I think this was meant to end t "same name." Followed by "Type-checking the function expression...".

File: lib/Sema/CSDiag.cpp [hunk 6]
```diff
@@ -6246,6 +6282,35 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
+  // expression separately from arguments might produce solution for
+  // the property instead of the member.
+  if (!fnType->is<AnyFunctionType>()) {
+    if (auto *UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn())) {
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: I don't see UDE being used. Can this be:
```
  if (!fnType->is<AnyFunctionType>() && isa<UnresolvedDotExpr>(callExpr->getFn())) {
```

File: lib/Sema/CSDiag.cpp [hunk 7]
```diff
@@ -6246,6 +6282,35 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
...
+      CS->TC.getPossibleTypesOfExpressionWithoutApplying(fnExpr, CS->DC, types);
+
+      unsigned numFuncTypes = 0;
+      Type funcType;
+      for (auto type : types) {
+        if (getFuncType(type)->is<AnyFunctionType>()) {
+          funcType = getFuncType(type);
+          numFuncTypes++;
+        }
+      }
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: This seems like a good place where `std::find_if` followed by `std::none_of` (with the range after the result of `find_if`) would make the code clearer.

**Comment by xedin** [comment]
@rudkx addressed all of your comments, thanks!

**Comment by xedin** [comment]
@swift-ci please smoke test

File: test/ClangImporter/objc_parse.swift [hunk 8]
```diff
@@ -310,7 +310,7 @@ class NSObjectable : NSObjectProtocol {
 // Properties with custom accessors
 func customAccessors(_ hive: Hive, bee: Bee) {
   markUsed(hive.isMakingHoney)
-  markUsed(hive.makingHoney()) // expected-error{{cannot call value of non-function type 'Bool'}}{{28-30=}}
+  markUsed(hive.makingHoney()) // expected-error{{cannot call value of non-function type '@lvalue Bool'}}{{28-30=}}
```
Comment (rudkx) [inline test/ClangImporter/objc_parse.swift]: I guess I missed this before, but we don't want to be printing `@lvalue` here.

Comment (xedin) [inline test/ClangImporter/objc_parse.swift]: Ah, sorry, it looks like it should be using `fnType` instead of `CS->getType(fnType)` in for that message, I'll change that.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by rudkx — APPROVED** [review]
